src/arq/_saw.py

```diff
@@ -0,0 +1,133 @@
+# -*- coding: utf-8 -*-
+
+
+import gevent.queue
+import itertools
+import random
+import struct
+
+from binascii import hexlify
+from enum import IntEnum
+
+from ._utils import concurrently
+
+
+class PacketType(IntEnum):
+    data = 0
+    ackn = 1
+
+
+HEADER = struct.Struct('>BH')
+HEADER_SIZE = HEADER.size
+
+
+def sequence_numbers():
+    """Generate sequence numbers that wrap around."""
+    for i in itertools.count(start=1):
+        yield i % 0xffff
+
+
+def recvfrom(socket, peer, size):
+    """Receive a packet from a designated peer."""
+    data, p = socket.recvfrom(size)
+    while p != peer:
+        data, p = socket.recvfrom(size)
+    return data
+
+
+def stop_and_wait(socket, peer, iq, oq, prefix='', p=0.0):
+    """."""
+
+    aq = gevent.queue.Queue()
+
+    def recv():
+        for i in sequence_numbers():
+            while True:
+                print(prefix + 'recv:', 'blocking')
+                data = recvfrom(socket, peer, 1024)
+                if random.random() < p:
+                    print(prefix + 'recv:', 'drop')
+                    continue
+                print(prefix + 'recv:', 'body=%r' % (
+                    hexlify(data),
+                ))
+                if len(data) < HEADER.size:
+                    print(prefix + 'recv:', 'packet too small (size=%d).' % (
+                        len(data),
+                    ))
+                t, j = HEADER.unpack(data[:HEADER.size])
+                if t == PacketType.data:
+                    if j == i:
+                        print(prefix + 'recv:', 'data packet #%d (accept)' % (
+                            j,
+                        ))
+                        socket.sendto(
+                            HEADER.pack(PacketType.ackn, i), peer,
+                        )
+                        iq.put(data[HEADER.size:])
```

Comment:
**TODO**: prevent blocking when `iq` is full (else we cannot forward incoming `ACKN` packets.

Reply:
Done.